### PR TITLE
feat: implement Platform Ban PoSe DIP-0031

### PR DESCRIPTION
**Pull request by knst**
## Issue being fixed or feature implemented
It implements [DIP-0031 ](https://github.com/dashpay/dips/blob/master/dip-0031.md)

## What was done?
Implemented p2p message in inventory system and functional tests for it.

## How Has This Been Tested?
See `test/functional/p2p_platform_ban.py`

## Breaking Changes
It bumps protocol version `MIN_MASTERNODE_PROTO_VERSION` for MN to 70236.

## Checklist:
- [ ] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

**Comment by UdjinM6** [comment]
tests are failing

**Comment by knst** [comment]
> tests are failing

yes, it is indeed failing due to more strict validations during mining quorum, introduced by #6645
Should succeed now

File: test/functional/test_framework/messages.py [hunk 1]
```diff
@@ -2345,6 +2345,41 @@ def __repr__(self):
...
+        self.requested_height= struct.unpack("<I", f.read(4))[0]
+        self.quorum_hash = deser_uint256(f)
+        self.sig = f.read(96)
+
+    def serialize(self):
+        r = b""
+        r += ser_uint256(self.protx_hash)
+        r += struct.pack("<I", self.requested_height)
+        r += ser_uint256(self.quorum_hash)
+        r += self.sig
```
[the remark this answers is not in the document]
Comment (knst) [reply test/functional/test_framework/messages.py:2369]: m_signature is named accordingly code style; the name m_signature is not matched with python's code style

**Comment by DashCoreAutoGuix** [comment]
Guix Automation has began to build this PR tagged as v23.0.0-devpr6613.b94d7255. A new comment will be made when the image is pushed.

**Comment by DashCoreAutoGuix** [comment]
Guix Automation has completed; a release should be present here: https://github.com/dashpay/dash-dev-branches/releases/tag/v23.0.0-devpr6613.b94d7255. The image should be on dockerhub soon.

**Comment by PastaPastaPasta** [comment]
@coderabbitai review

**Comment by PastaPastaPasta** [comment]
Consider:
  Areas for Improvement:

  1. Magic Numbers: The 576 block range check uses a hardcoded value:
  const int day_of_blocks = 576;
  1. Consider moving this to consensus parameters or defining as a constant.
  2. Error Handling: Different penalty values (1, 10, 100) for various validation failures could be better documented:
  return tl::unexpected{1};   // Recent removal
  return tl::unexpected{10};  // Height out of range
  return tl::unexpected{100}; // Invalid node type or signature
  3. Memory Management: The m_seen_platform_bans map grows indefinitely. Consider implementing a size limit or TTL mechanism.
  4. Race Conditions: In SetPlatformBan, the height comparison logic could be more explicit:
  if (height == m_platform_ban_height && !is_banned) {
      return false;  // Comment explaining why unbanning at same height is rejected
  }

**Comment by PastaPastaPasta** [comment]
@coderabbitai, how well tested is this PR? Are there tests that could / or should be written for the features introduced here?

**Comment by knst** [comment]
REBASE COUNTER (due to conflicts): 6

**Comment by knst** [comment]
@PastaPastaPasta @UdjinM6 please help with review this PR and get it merged

File: src/evo/deterministicmns.cpp [hunk 2]
```diff
@@ -641,6 +642,23 @@ bool CDeterministicMNManager::ProcessBlock(const CBlock& block, gsl::not_null<co
...
+            const auto opt_proTx = GetTxPayload<CProUpServTx>(tx);
+            if (!opt_proTx) continue; // should not happen but does not matter
+
+            if (auto meta_info = m_mn_metaman.GetMetaInfo(opt_proTx->proTxHash, false);
+                !meta_info || !meta_info->SetPlatformBan(false, nHeight)) {
+                LogPrint(BCLog::LLMQ, "%s -- MN %s is failed to Platform revived at height %d\n", __func__,
+                         opt_proTx->proTxHash.ToString(), nHeight);
+            }
+        }
+
```
Comment (UdjinM6) [inline src/evo/deterministicmns.cpp:661]: what about UndoBlock()?

File: src/masternode/meta.h [hunk 3]
```diff
@@ -46,6 +48,9 @@ class CMasternodeMetaInfo
     std::atomic<int64_t> lastOutboundAttempt{0};
     std::atomic<int64_t> lastOutboundSuccess{0};
 
+    bool m_platform_ban GUARDED_BY(cs){false};
+    int m_platform_ban_height GUARDED_BY(cs){0};
```
Comment (UdjinM6) [inline src/masternode/meta.h]: do we really care when a node was _unbanned_? could use `m_platform_ban_height == 0` as `not banned`, drop `m_platform_ban` and simplify `SetPlatformBan`

File: src/masternode/meta.h [hunk 4]
```diff
@@ -55,22 +60,18 @@ class CMasternodeMetaInfo
...
-                obj.nLastDsq,
-                obj.nMixingTxCount,
-                obj.mapGovernanceObjectsVotedOn,
-                obj.outboundAttemptCount,
-                obj.lastOutboundAttempt,
-                obj.lastOutboundSuccess
-                );
+        READWRITE(obj.proTxHash, obj.nLastDsq, obj.nMixingTxCount, obj.mapGovernanceObjectsVotedOn,
+                  obj.outboundAttemptCount, obj.lastOutboundAttempt, obj.lastOutboundSuccess, obj.m_platform_ban,
+                  obj.m_platform_ban_height);
```
Comment (UdjinM6) [inline src/masternode/meta.h]: should bump `MasternodeMetaStore::SERIALIZATION_VERSION_STRING`

File: src/masternode/meta.h [hunk 5]
```diff
@@ -150,6 +169,40 @@ class MasternodeMetaStore
...
+    uint256 m_quorum_hash;
+    CBLSSignature m_signature;
+
+    PlatformBanMessage() = default;
+
+    SERIALIZE_METHODS(PlatformBanMessage, obj)
+    {
+        READWRITE(obj.m_protx_hash, obj.m_requested_height, obj.m_quorum_hash);
+        if (!(s.GetType() & SER_GETHASH)) {
+            READWRITE(CBLSSignatureVersionWrapper(const_cast<CBLSSignature&>(obj.m_signature), false));
```
Comment (UdjinM6) [inline src/masternode/meta.h:204]: do we really need a wrapper here?

File: src/evo/deterministicmns.cpp (hunk 2 above)
Comment (knst) [reply src/evo/deterministicmns.cpp:661]: if masternode already got PoSe banned, it's not a problem, we don't care anymore.
If masternode has not got PoSe banned yet, but we have seen protx to revive it - we may not vote anymore for this node to be banned for awhile, but apparently new message of `platform ban` will be send to network if the node still doesn't provide evo service.

Also, I think it's a rare situation, when protx got to the mined block and chain switched to new tip without it and this protx never appeared again.

There's still possible situation, when block can be `undo` manually, by calling `invalidateblock`, but it should not affect overall "banned / not banned" condition, because it's used only for voting.

I think implementing similar logic for `UndoBlock` will make implementation of platform-ban less reliable rather than improve it.

File: src/masternode/meta.h (hunk 3 above)
Comment (knst) [reply src/masternode/meta.h]: yes, we really care when a node has been banned last time by p2p message.
If node has been revived and after that we received p2p message about ban we should not vote for banning it again.
That's why we should not change `m_platform_ban_height` after unbanning

File: src/masternode/meta.h (hunk 4 above)
Comment (knst) [reply src/masternode/meta.h]:
```
s >> tmpMetaInfo >> nDsqCount;
```
yes, indeed, if the version is changed, nDsgCount can be interpreted as a part of metaInfo (due to changed size of structure).

File: src/masternode/meta.h (hunk 5 above)
Comment (knst) [reply src/masternode/meta.h:204]: yes, because I want to enforce basic BLS scheme here; I assume it can happen only in theory during re-index when it would be serialized as legacy, but let enforce basic better.

File: src/masternode/meta.cpp [hunk 6]
```diff
@@ -125,8 +130,31 @@ std::vector<uint256> CMasternodeMetaMan::GetAndClearDirtyGovernanceObjectHashes(
...
+    if (it == m_seen_platform_bans.end()) return std::nullopt;
+
+    return it->second;
+}
+
+void CMasternodeMetaMan::RememberPlatformBan(const uint256& inv_hash, PlatformBanMessage& msg)
+{
+    LOCK(cs);
+    m_seen_platform_bans.insert({inv_hash, msg});
+}
```
[the remark this answers is not in the document]
Comment (UdjinM6) [reply src/masternode/meta.cpp]: Not sure about lru but making `msg` const makes sense imo

[the remark this answers is not in the document]
Comment (knst) [reply src/masternode/meta.cpp]: > Not sure about lru but making msg const makes sense imo

I think lru is reasonable here to avoid potential DoS by RAM overflow if L2 will find a way to spam messages somehow. Though, instead `const` I used '&&' here to avoid copy of heavy object with Bls Signature.

File: src/masternode/meta.h [hunk 7]
```diff
@@ -150,6 +169,40 @@ class MasternodeMetaStore
...
 };
 
+/**
+ * Platform PoSe Ban are result in the node voting against the targeted evonode in all future DKG sessions until that targeted
+ *evonode has been successfully banned. Platform will initiate this ban process by passing relevant information to Core using RPC. See DIP-0031
+ *
+ * We use 2 main classes to manage Platform PoSe Ban
+ *
+ * PlatformBanMessage
+ * PlatformBanManager - a higher-level construct which store information about ban status
```
Comment (UdjinM6) [inline src/masternode/meta.h]: There is no `PlatformBanManager`... should maybe create one in a new file and move `PeerManagerImpl::ProcessPlatformBanMessage` logic into this new class. Could also move  `PlatformBanMessage` into that file.

Comment (knst) [reply src/masternode/meta.h]:
```
-        ret.pushKV("platform_ban_updated", m_platform_ban_height);
+        ret.pushKV("platform_ban_height", m_platform_ban_height);
```

platform_ban_updated is more correct name; because unbanning node (by protx) updates it too. I will rename member variable accordingly.


> You’ve correctly used an rvalue overload to avoid copies, but callers holding an lvalue must still use std::move. For ergonomics, add a const& overload.

I added '&&' here intentionally to avoid copying object, but I can revert this change 🤷

---
I updated doc, commit is comming....

File: src/net_processing.cpp [hunk 8]
```diff
@@ -723,6 +724,9 @@ class PeerManagerImpl final : public PeerManager
                                const std::vector<CBlockHeader>& headers,
                                bool via_compact_block)
         EXCLUSIVE_LOCKS_REQUIRED(!m_peer_mutex, g_msgproc_mutex);
+    [[nodiscard]] MessageProcessingResult ProcessPlatformBanMessage(NodeId from, std::string_view msg_type, CDataStream& vRecv)
```
Comment (UdjinM6) [inline src/net_processing.cpp]:
```suggestion
    [[nodiscard]] MessageProcessingResult ProcessPlatformBanMessage(NodeId node, std::string_view msg_type, CDataStream& vRecv)
```

File: src/net_processing.cpp [hunk 9]
```diff
@@ -3507,6 +3520,81 @@ void PeerManagerImpl::PostProcessMessage(MessageProcessingResult&& result, NodeI
     }
 }
 
+MessageProcessingResult PeerManagerImpl::ProcessPlatformBanMessage(NodeId from, std::string_view msg_type, CDataStream& vRecv)
```
Comment (UdjinM6) [inline src/net_processing.cpp]:
```suggestion
MessageProcessingResult PeerManagerImpl::ProcessPlatformBanMessage(NodeId node, std::string_view msg_type, CDataStream& vRecv)
```

File: src/net_processing.cpp [hunk 10]
```diff
@@ -3507,6 +3520,81 @@ void PeerManagerImpl::PostProcessMessage(MessageProcessingResult&& result, NodeI
...
+    if (!m_mn_sync.IsBlockchainSynced()) {
+        return {};
+    }
+
+    PlatformBanMessage ban_msg;
+    vRecv >> ban_msg;
+
+    const uint256 hash = ban_msg.GetHash();
+
+    LogPrintf("PLATFORMBAN -- hash: %s protx_hash: %s height: %d peer=%d\n", hash.ToString(), ban_msg.m_protx_hash.ToString(), ban_msg.m_requested_height, from);
```
Comment (UdjinM6) [inline src/net_processing.cpp]:
```suggestion
    LogPrintf("PLATFORMBAN -- hash: %s protx_hash: %s height: %d peer=%d\n", hash.ToString(), ban_msg.m_protx_hash.ToString(), ban_msg.m_requested_height, node);
```

File: src/net_processing.cpp (hunk 8 above)
Comment (UdjinM6) [reply src/net_processing.cpp]: @knst you missed this one

**Comment by knst** [comment]
> 727-729: Correct signature and attributes for ProcessPlatformBanMessage
> Nice: [[nodiscard]] and the locking annotations are consistent with the surrounding handlers.

no comments, thank you

> 3553-3560: Avoid magic number 576: use a named constant for the window

There's already named constant
```cpp
    constexpr int day_of_blocks = 576;
```

Independently, same literal 576 is used to clean "day old quorum data", 576 is used as CDeterministicMNList::DISK_SNAPSHOT_PERIOD for CCreditPool::DISK_SNAPSHOT_PERIOD as a market to create a snapshot. But all of them are independent and there should not be extra dependencies between various components just because "human-friendly" 1 day constant. I'd prefer to keep as it is or rename `day_of_blocks` to make code self-documented.



> 3537-3546: Add message to MisbehavingError for better diagnostics
There are relevant logs for each MisbehavingError case:
```
LogPrintf("PLATFORMBAN -- hash: %s protx_hash: %s unexpected type of node\n", hash.ToString(), ban_msg.m_protx_hash.ToString());
LogPrintf("PLATFORMBAN -- hash: %s protx_hash: %s unexpected height: %d tip: %d\n", hash.ToString(), ban_msg.m_protx_hash.ToString(), ban_msg.m_requested_height, tipHeight);
LogPrintf("PLATFORMBAN -- hash: %s protx_hash: %s missing quorum_hash: %s llmq_type: %d\n", hash.ToString(), ban_msg.m_protx_hash.ToString(), ban_msg.m_quorum_hash.ToString(), ToUnderlying(llmq_type));
...
```

> 3556-3560: Clarify comment about acceptable height range
Wording is unified with DIP-0031

File: src/net_processing.cpp [hunk 11]
```diff
@@ -3507,6 +3520,83 @@ void PeerManagerImpl::PostProcessMessage(MessageProcessingResult&& result, NodeI
...
+    PlatformBanMessage ban_msg;
+    vRecv >> ban_msg;
+
+    const uint256 hash = ban_msg.GetHash();
+
+    LogPrintf("PLATFORMBAN -- hash: %s protx_hash: %s height: %d peer=%d\n", hash.ToString(), ban_msg.m_protx_hash.ToString(), ban_msg.m_requested_height, node);
+
+    WITH_LOCK(::cs_main, EraseObjectRequest(node, CInv{MSG_PLATFORM_BAN, hash}));
+
+    MessageProcessingResult ret{};
```
Comment (UdjinM6) [inline src/net_processing.cpp]: I htink we should let `PostProcessMessage` do the job
```suggestion
    MessageProcessingResult ret{};
    ret.m_to_erase = CInv{MSG_PLATFORM_BAN, hash};

```

File: src/net_processing.cpp [hunk 12]
```diff
@@ -3507,6 +3520,83 @@ void PeerManagerImpl::PostProcessMessage(MessageProcessingResult&& result, NodeI
...
+    const auto meta_info = m_mn_metaman.GetMetaInfo(ban_msg.m_protx_hash);
+    if (meta_info->SetPlatformBan(true, ban_msg.m_requested_height)) {
+        LogPrintf("PLATFORMBAN -- forward message to other nodes\n");
+        m_mn_metaman.RememberPlatformBan(hash, std::move(ban_msg));
+        CInv platform_ban_inv{MSG_PLATFORM_BAN, hash};
+        RelayInv(platform_ban_inv);
+    }
+    return {};
+}
+
```
[the remark this answers is not in the document]
Comment (UdjinM6) [reply src/net_processing.cpp:3616]: Good catch! :+1:

File: src/net_processing.cpp [hunk 13]
```diff
@@ -3507,6 +3520,83 @@ void PeerManagerImpl::PostProcessMessage(MessageProcessingResult&& result, NodeI
...
+        return ret;
+    }
+
+    // At this point, the outgoing message serialization version can't change.
+    const auto meta_info = m_mn_metaman.GetMetaInfo(ban_msg.m_protx_hash);
+    if (meta_info->SetPlatformBan(true, ban_msg.m_requested_height)) {
+        LogPrintf("PLATFORMBAN -- forward message to other nodes\n");
+        m_mn_metaman.RememberPlatformBan(hash, std::move(ban_msg));
+        CInv platform_ban_inv{MSG_PLATFORM_BAN, hash};
+        RelayInv(platform_ban_inv);
```
Comment (UdjinM6) [inline src/net_processing.cpp]: This part can also be delegated to `PostProcessMessage`
```suggestion
        ret.m_inventory = CInv{MSG_PLATFORM_BAN, hash};
```

File: src/net_processing.cpp [hunk 14]
```diff
@@ -3507,6 +3520,83 @@ void PeerManagerImpl::PostProcessMessage(MessageProcessingResult&& result, NodeI
...
+    }
+    return {};
+}
+
```
[the remark this answers is not in the document]
Comment (knst) [reply src/net_processing.cpp:3616]: indeed! Fix is coming

File: src/net_processing.cpp [hunk 15]
```diff
@@ -3591,7 +3591,7 @@ MessageProcessingResult PeerManagerImpl::ProcessPlatformBanMessage(NodeId node,
     if (meta_info->SetPlatformBan(true, ban_msg.m_requested_height)) {
         LogPrintf("PLATFORMBAN -- forward message to other nodes\n");
         m_mn_metaman.RememberPlatformBan(hash, std::move(ban_msg));
-        ret.m_inventory = CInv{MSG_PLATFORM_BAN, hash};
+        ret.m_inventory.emplace_back(MSG_PLATFORM_BAN, hash);
```
Comment (UdjinM6) [inline src/net_processing.cpp:3612]: oops, my bad 🙈

File: src/masternode/meta.h [hunk 16]
```diff
@@ -96,6 +101,24 @@ class CMasternodeMetaInfo
     int64_t GetLastOutboundAttempt() const { return lastOutboundAttempt; }
     void SetLastOutboundSuccess(int64_t t) { lastOutboundSuccess = t; outboundAttemptCount = 0; }
     int64_t GetLastOutboundSuccess() const { return lastOutboundSuccess; }
+    bool SetPlatformBan(bool is_banned, int height)
+    {
+        LOCK(cs);
+        if (height <= m_platform_ban_updated) {
+            return false;
+        }
+        if (height == m_platform_ban_updated && !is_banned) {
```
Comment (UdjinM6) [inline src/masternode/meta.h:110]: This condition is unreachable now because we'll bail out in `if (height <= m_platform_ban_updated) {` above (it was changed from `if (height < m_platform_ban_height) {` in https://github.com/dashpay/dash/pull/6613/commits/c739f3a6e001ce1a8928d81fd8bf0772d45e5d3e#diff-4b932ee0759e8f52676ccdd2f94fb3a39ee0652695edfa3822ecb3dcb7ef9f30R107)

File: src/net_processing.cpp [hunk 17]
```diff
@@ -3507,6 +3520,83 @@ void PeerManagerImpl::PostProcessMessage(MessageProcessingResult&& result, NodeI
...
+    static constexpr int PLATFORM_BAN_WINDOW_BLOCKS = 576;
+    int tipHeight = WITH_LOCK(cs_main, return m_chainman.ActiveChainstate().m_chain.Height());
+    if (tipHeight < ban_msg.m_requested_height || tipHeight - PLATFORM_BAN_WINDOW_BLOCKS > ban_msg.m_requested_height) {
+        // m_requested_height is inside the range [TipHeight - PLATFORM_BAN_WINDOW_BLOCKS - 5, TipHeight + 5]
+        LogPrintf("PLATFORMBAN -- hash: %s protx_hash: %s unexpected height: %d tip: %d\n", hash.ToString(), ban_msg.m_protx_hash.ToString(), ban_msg.m_requested_height, tipHeight);
+        if (tipHeight + 5 < ban_msg.m_requested_height || tipHeight - PLATFORM_BAN_WINDOW_BLOCKS - 5 > ban_msg.m_requested_height) {
+            // m_requested_height is outside the range [TipHeight - PLATFORM_BAN_WINDOW_BLOCKS - 5, TipHeight + 5]
+            ret.m_error = MisbehavingError{10};
+            return ret;
+        }
```
Comment (UdjinM6) [inline src/net_processing.cpp:3581]: `        // m_requested_height is inside the range [TipHeight - PLATFORM_BAN_WINDOW_BLOCKS - 5, TipHeight + 5]` should be moved lower, we don't know it at this moment yet (it's the implicit `else` part of the `if` below
```suggestion
        LogPrintf("PLATFORMBAN -- hash: %s protx_hash: %s unexpected height: %d tip: %d\n", hash.ToString(), ban_msg.m_protx_hash.ToString(), ban_msg.m_requested_height, tipHeight);
        if (tipHeight + 5 < ban_msg.m_requested_height || tipHeight - PLATFORM_BAN_WINDOW_BLOCKS - 5 > ban_msg.m_requested_height) {
            // m_requested_height is outside the range [TipHeight - PLATFORM_BAN_WINDOW_BLOCKS - 5, TipHeight + 5]
            ret.m_error = MisbehavingError{10};
            return ret;
        }
        // m_requested_height is inside the range [TipHeight - PLATFORM_BAN_WINDOW_BLOCKS - 5, TipHeight + 5]
```

File: test/functional/p2p_platform_ban.py [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+
+        recsig = self.get_recovered_sig(request_id, msg_hash, llmq_type=106, use_platformsign=True)
+        msg.sig = bytearray.fromhex(recsig["sig"])
+
+        self.log.info("Platform ban message is created and signed")
+        p2p_node2 = self.mninfo[2].get_node(self).add_p2p_connection(PlatformBanInterface())
+        p2p_node2.send_message(msg)
+        wait_until_helper(lambda: p2p_node2.message_count["platformban"] > 0, timeout=10, lock=p2p_lock)
+        p2p_node2.message_count["platformban"] = 0
+
```
Comment (UdjinM6) [inline test/functional/p2p_platform_ban.py:68]:
```suggestion
```

File: test/functional/p2p_platform_ban.py [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+        self.mine_quorum(llmq_type_name='llmq_test_platform', expected_members=2, expected_connections=1, expected_contributions=2, expected_commitments=2, llmq_type=106, mninfos_valid=mninfos_valid)
+
+        p2p_node = node.add_p2p_connection(PlatformBanInterface())
+        p2p_node.send_message(msg)
+        self.mine_quorum(llmq_type_name='llmq_test_platform', expected_members=2, expected_connections=1, expected_contributions=2, expected_commitments=2, llmq_type=106, mninfos_valid=mninfos_valid)
+        assert self.check_banned(self.mninfo[0])
+
+        self.dynamically_evo_update_service(evo_info_0)
+        assert not self.check_banned(self.mninfo[0])
+
```
Comment (UdjinM6) [inline test/functional/p2p_platform_ban.py:81]:
```suggestion


```

File: test/functional/p2p_platform_ban.py [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+        p2p_node2.message_count["platformban"] = 0
+
+
+        assert not self.check_banned(self.mninfo[0])
+
+        mninfos_valid = self.mninfo.copy()[1:]
+        self.mine_quorum(llmq_type_name='llmq_test_platform', expected_members=2, expected_connections=1, expected_contributions=2, expected_commitments=2, llmq_type=106, mninfos_valid=mninfos_valid)
+
+        p2p_node = node.add_p2p_connection(PlatformBanInterface())
+        p2p_node.send_message(msg)
```
Comment (UdjinM6) [inline test/functional/p2p_platform_ban.py:75]: why do we have to send it to the controller node (node0)?

**Comment by PastaPastaPasta** [comment]
@coderabbitai review

File: src/masternode/meta.h (hunk 16 above)
Comment (UdjinM6) [reply src/masternode/meta.h:110]: This is still confusing to me tbh... So this condition is basically "evonode CAN be banned at the height it was unbanned but it CAN NOT be unbanned at the height it was banned", right? We need a test for this edge case I guess.

EDIT:
In other words, if we process a block with `protxupserv` and then we receive `platformban` for the same height - we mark the node as __bad__. But if we receive `platformban` from the future (and mark the node as __bad__) first and we process a block with `protxupserv` for the same height later - we keep the node as __bad__.

We do need to update dip-0031 accordingly imo.

Comment (knst) [reply src/masternode/meta.h:110]: I am not sure why exactly I implemented this one like that, I had plan to update DIP-0031 accordingly implementation, but now I can't remember any more what is exactly not the same as DIP-0031 so far as half year passed since I implemented it.

I definitely want to prevent flipping ban/unban status. I added one more commit which change logic a bit. Could you have a look to it?

File: src/net_processing.cpp (hunk 17 above)
Comment (knst) [reply src/net_processing.cpp:3581]: > we don't know it at this moment yet

we *know* that it outside of range. But we don't know how far is it, that's why 2nd `if` is here.
So, first "if" checks if we are outside of range at all.
2nd "if" - to understand if we are far or close?

File: test/functional/p2p_platform_ban.py (hunk 20 above)
Comment (knst) [reply test/functional/p2p_platform_ban.py:75]: to be sure that this node will forward p2p message to other nodes

File: src/masternode/meta.h [hunk 21]
```diff
@@ -104,10 +104,7 @@ class CMasternodeMetaInfo
     bool SetPlatformBan(bool is_banned, int height)
     {
         LOCK(cs);
-        if (height <= m_platform_ban_updated) {
-            return false;
-        }
-        if (height == m_platform_ban_updated && !is_banned) {
+        if (height < m_platform_ban_updated) {
```
Comment (UdjinM6) [inline src/masternode/meta.h:107]: >no flipping ban/unban if it's the same height

but without `=` it _will_ be flipping for the same height...

Comment (knst) [reply src/masternode/meta.h:107]: yes, right, indeed.

I updated condition one more time: I tried to recover original logic:
 - it should not flip
 - all nodes should have the same decision, if they see block with the same height and p2p message with the same height, not 50-50 depending what mn see earlier - p2p message or a block

**Review by UdjinM6 — APPROVED** [review]
LGTM, utACK 4fc84c7c639d4b91930601aced29478ea164248e

File: src/masternode/meta.h [hunk 22]
```diff
@@ -55,22 +64,18 @@ class CMasternodeMetaInfo
...
-                obj.nMixingTxCount,
-                obj.mapGovernanceObjectsVotedOn,
-                obj.outboundAttemptCount,
-                obj.lastOutboundAttempt,
-                obj.lastOutboundSuccess
-                );
+        READWRITE(obj.proTxHash, obj.nLastDsq, obj.nMixingTxCount, obj.mapGovernanceObjectsVotedOn,
+                  obj.outboundAttemptCount, obj.lastOutboundAttempt, obj.lastOutboundSuccess, obj.m_platform_ban,
+                  obj.m_platform_ban_updated);
     }
```
[the remark this answers is not in the document]
Comment (UdjinM6) [reply src/masternode/meta.h:79]: we bumped it from 3 to 4 in this PR, no need to bump any further

**Review by PastaPastaPasta — APPROVED** [review]
utACK 64d2f67d4b6c2fb78efc1f4f530d936df55e867c
